### ISSUE TITLE
Rendering errors for stateDiagram-2

### ISSUE THREAD

**Issue by sjn** [mermaid-js/mermaid#5827]
### Description

I'm working on (and have published) a couple of stateDiagrams that seem to be broken with the v11.1.0 renderer.

**Found:**

You can find them "live" here: https://github.com/CPAN-Security/security.metacpan.org/blob/supplychain-sbom/docs/supplychain-sbom.md (using v11.1.0 as of 2022-09-06)

**Expected:**

This diagram differs in many ways from the "live" one, which uses the v10.0.9 renderer: https://security.metacpan.org/docs/supplychain-sbom.html#open-source-supply-chain-full-diagram

**Regressions:**

Apologies for using my own terms here; I hope you can still use this to figure out what I mean

- [ ] The TB (top-bottom) bias has been removed, making state blocks render more often "besides" each other, instead of below
- [ ] Self-referring links from/to state blocks (groups) look "boxy" compared to self-referring links to regular states
- [ ] Self-referring links from/to state blocks/groups don't always take the "shortest route" to itself.
- [ ] State description text wrap based on the size of the available area, instead of the previous behavior, where the area expanded based on the text length
- [ ] State description doesn't respect newlines any more
- [ ] State description text is centered instead of left-adjusted as previously

### Steps to reproduce

Please see the guthub source of the pages linked above: https://github.com/CPAN-Security/security.metacpan.org/blob/supplychain-sbom/docs/supplychain-sbom.md?plain=1#L203-L378

### Screenshots

Expected:

![image](https://github.com/user-attachments/assets/e519d6a0-9d8e-4792-beae-571fcb5af0a9)

Found:

![image](https://github.com/user-attachments/assets/890f91ef-8f00-4926-9c2e-8fef7cdf4705)
![image](https://github.com/user-attachments/assets/310722c6-48a3-43d1-8d06-46f08a3fca25)

### Setup

- Mermaid version: Whatever github is using (I'm told it's 11.1.0)
- Browser and Version: Firefox 127.0 for Ubuntu Linux

**Comment by sjn** [comment]
BTW, when I say the self-referring links look "boxy", I'm comparing it to these:

![image](https://github.com/user-attachments/assets/4de22d6a-ccb6-4f9a-ba29-ee0c8c977980)

They look a LOT nicer, and the only grip I have with them is that in a top-bottom direction biased diagram, I expect the "exits" to be at the bottom of the node, and the "entrances" be on the top – meaning that the directionality if the arrow has a bias/preference for going from the bottom to the top _including when the arrows are self-referring_.

I hope that makes sense? :-)

(I've updated the OP with this)

**Comment by knsv** [comment]
@sjn Thanks for reporting. We are publishing a hotfix as soon as possible. I have checked that your diagram will render after the hotfix. I agree that the self-loops look a bit boxy, and we will try to fix them in a future release. 

Relates #5820, #5822

**Comment by sjn** [comment]
Is there a chance that any of the other rendering buts will be addressed soon? I'm noticing that more people are upgrading, which makes this a bit more difficult for me who would like to be able to present clear flowcharts (without strange rendering oddities).

Is there anything I can do to assist?

**Comment by waimea-cpy** [comment]
Just to chip in with some other examples...

I've had to revert to using v10 again, since the self-referencing loops are just too janky.

I'm trying to create Finite State Automata diagrams. This graph...

```
flowchart LR

S0[ ]
S1((S1))
S2(((S2)))
S3(((S3)))

S0 --> S1
S1 --A--> S2
S1 --B--> S3
S2 --A--> S1
S2 --B--> S3
S3 --A--> S1
S3 --B--> S3
S3 --C--> S3

class S0 start
```

... looks kind of ok using v10 (the overlapping loops aren't ideal, but at least they are readable):

![Screenshot 2024-11-13 092023](https://github.com/user-attachments/assets/15e4b537-0e53-4fa5-9c15-7ad82facdd18)

... but with v11, the self-referencing links are rendered with a sharp point, but more importantly, **one covers the other**:

![Screenshot 2024-11-13 092133](https://github.com/user-attachments/assets/a324d4be-c9e3-4f34-8ea4-a05431ae7cb0)

**Comment by scudelletti** [comment]
@waimea-cpy have you noticed that one of the arrows(B) from S3 to S3 disappeared? Do you know if that's expected?

That happens on v11 but works fine on v10.
Issue #6034 also describes issues with arrows in the latest version, but on class diagrams.

Example in Flowchart:
```
flowchart TD
  a-->|1| b
  a-->|2| b
  a-->|3| a
  a-->|4| a
```

```mermaid
flowchart TD
  a-->|1| b
  a-->|2| b
  a-->|3| a
  a-->|4| a
```

v11:
![a_new-1](https://github.com/user-attachments/assets/ff7435e0-215c-4a8c-ad4f-3ae606320d24)

v10:
![a_old-1](https://github.com/user-attachments/assets/6b30a90b-f083-4029-b7ed-45a5c956fe50)

**Comment by waimea-cpy** [comment]
Yes, you can see the same issue in my examples... One arrow is obscured by another. I mentioned this in the comment.

However, this only occurs in v11.1 onwards. v11.0.2 doesn't have this bug.

**Comment by hoijnet** [comment]
I have not seen in the referenced issues any mention of this also affecting `classDiagram`, but they have the same issue too. For flowcharts I could use the workaround with selecting the elk engine.

I could not understand how to add a configuration in Mermaid Live for setting the default engine to elk which hopefully can work for now until fixed.

Applying below in Mermaid Live with latest version v11.4.1 yields the same issue below.

```
classDiagram
  `SelfReference` --> `SelfReference` : parent (L)
  class `SelfReference` {
    parent ＜SelfReference＞
  }
```

<img width="668" alt="Image" src="https://github.com/user-attachments/assets/c6450b15-634e-4bba-b46a-e2ea14080739" />

Thanks for raising the issues and looking into it. If anyone knows how to add a configuration pre-amble that can set the default renderer, it might be possible to do a workaround with the elk renderer as with for example the flowchart.
